accepted/2024/release-staging-and-deprecation.md

```diff
@@ -0,0 +1,193 @@
+- **Stage** a package for later publishing
+- **Publish an earlier staged group**
+- **Deprecate a group** using the group id 
+- **Discard a staged release** using the group id
+
+Note, that the user is required to be the owner / co-owner of the packages which are in the group,
+otherwise the request to make any modification to a non-owned package will fail.
+
+#### Package staging
+When a package is staged, the id of the package is reserved at that point, and no other user can introduce a package with the same name.
+
+Also, to avoid abuse, only a limited number of packages can be staged within a single group. While this can potentially be defined as a configurable value per organization / owner, there is no such requirement at the moment. Hence, this can be set to 1000 packages for now or not defined at all. Will leave this for potentially future consideration.
+
+When staged, packages will not be visible / accessible to anyone but the co-owners. The co-owners will be able to see the current staged list for a given group using `nuget list --stage [group-id]` command. 
+
+
+#### Group Ownership
+There are few interesting scenarios related to package ownership:
+1. The user has staged a package to a group for later publishing, but the ownership has changed, and the user is no longer
+ an owner for a package in a group. At this point, publishing request should fail with a message that the group has been modified
+ from its original set, and the outcome of the request may be different from what the user was expecting, indicating the package
+ that was moved out of the group. If the user is aware of the problem and still wants to move forward with the change,
+ they will have to pass an additional `--force` parameter to the publish command: `nuget push --stage <group-id> --force`.
+2. The group to be published no longer has any packages in it. In this case the command should fail detailing the reason for the failure.
+3. This will be the same but for deprecation. Essentially, any bulk operation using groups, where the set of the packages in the group
+ have changed after staging, should result in an error with details about the change. The `--force` flag should again be used
+ to ignore the change and move forward with execution of the request (either publishing or deprecation). To summarize,
+ the fundamental principle is that a user cannot deprecate a package they don't own.
+
+#### Lifespan of a staged group
+If users keep staging packages but do nothing with them, NuGet will become an ever-growing package graveyard,
+ incurring higher and higher costs over time simply for storing these packages. To avoid this problem, stages should be time-limited.
+ That is, there should be a maximum lifespan for a stage. Any updates to the stage will reset its TTL date to some amount (let's say 30 days). If within the next 30 days no changes are made to the stage, NuGet should then remove the stage (all its packages basically).
+
+#### Expanding groups
```

Comment:
This almost sounds like "group publishing" and "group deprecation" functionalities should be separate from each other.

Reply:
Can you please clarify what you mean, as these two indeed two separate functionalities.

Reply:
I don't see a connection between staged publishing and group deprecation and see no reason for those to be covered in a single proposal other than it might be useful for releasing .net specifically. I would really love to see some input here from somebody else.

Reply:
I agree with @agr. It's not clear to me that this bulk deprecation experience would generalize well to projects other than the .NET team.

Reply:
While this proposal covers both, they are indeed somewhat separate functionalities, not tied to each other. My personal motivation was to have the deprecation functionality, and it requires the staging / grouping to be in place to work. Hence the reason they both are included in the same proposal.